packages/d2mini/src/utils.ts

```diff
@@ -111,3 +111,53 @@ export function binarySearch<T>(
   }
   return low
 }
+
+/**
+ * Utility for generating unique IDs for objects and values.
+ * Uses WeakMap for object reference tracking and consistent hashing for primitives.
+ */
+export class ObjectIdGenerator {
+  private objectIds = new WeakMap<object, number>()
+  private nextId = 0
+
+  /**
+   * Get a unique identifier for any value.
+   * - Objects: Uses WeakMap for reference-based identity
+   * - Primitives: Uses consistent string-based hashing
+   */
+  getId(value: any): number {
+    // For primitives, use a simple hash of their string representation
+    if (typeof value !== 'object' || value === null) {
+      const str = String(value)
+      let hash = 0
+      for (let i = 0; i < str.length; i++) {
+        const char = str.charCodeAt(i)
+        hash = (hash << 5) - hash + char
+        hash = hash & hash // Convert to 32-bit integer
+      }
+      return hash
+    }
+
+    // For objects, use WeakMap to assign unique IDs
+    if (!this.objectIds.has(value)) {
+      this.objectIds.set(value, this.nextId++)
+    }
+    return this.objectIds.get(value)!
+  }
+
+  /**
+   * Get a string representation of the ID for use in composite keys.
+   */
+  getStringId(value: any): string {
+    if (value === null) return 'null'
+    if (value === undefined) return 'undefined'
+    if (typeof value !== 'object') return String(value)
+
+    return `obj_${this.getId(value)}`
```

Comment:
This object representation may clash with string values if the string would be `"obj_..."`.
Also, it seems that `null` and `undefined` would clash with the strings `"null"` and `"undefined"`.
We should avoid these problems by tagging the strings. So, every object starts with `obj_` and every string should start with `string_` that way we avoid clashes between strings and objects and between strings and null/undefined.